Description: Python 3.13: crypt.crypt support is dropped
 The module crypt is dropped from Python 3.13 and therefore, we
 must replace this by something else.
 .
 Also, the old code from 2012 was still using MD5, which is
 nowadays considered a security hole, because it is extremely
 easy to brute-force. So this patch is switching to SHA-512,
 dropping support for distros that don't support that (though
 it would be not reasonable to use such a distro).
Author: Thomas Goirand <zigo@debian.org>
Date: Mon, 18 Nov 2024 14:13:53 +0100
Change-Id: I6b5e9282806961180a11a7f0b0607233bf4fd700
Forwarded: https://review.opendev.org/c/openstack/nova/+/935512
Last-Update: 2024-11-19

Index: nova/nova/tests/unit/virt/disk/test_inject.py
===================================================================
--- nova.orig/nova/tests/unit/virt/disk/test_inject.py
+++ nova/nova/tests/unit/virt/disk/test_inject.py
@@ -206,28 +206,18 @@ class VirtDiskTest(test.NoDBTestCase):
                           'uid': 100})
         shadow = vfs.handle.files["/etc/shadow"]
 
-        # if the encrypted password is only 13 characters long, then
-        # nova.virt.disk.api:_set_password fell back to DES.
-        if len(shadow['content']) == 91:
-            self.assertEqual(shadow,
-                             {'content': "root:12tir.zIbWQ3c" +
-                                         ":14917:0:99999:7:::\n" +
-                                         "bin:*:14495:0:99999:7:::\n" +
-                                         "daemon:*:14495:0:99999:7:::\n",
-                              'gid': 100,
-                              'isdir': False,
-                              'mode': 0o700,
-                              'uid': 100})
-        else:
-            self.assertEqual(shadow,
-                             {'content': "root:$1$12345678$a4ge4d5iJ5vw" +
-                                         "vbFS88TEN0:14917:0:99999:7:::\n" +
-                                         "bin:*:14495:0:99999:7:::\n" +
-                                         "daemon:*:14495:0:99999:7:::\n",
-                              'gid': 100,
-                              'isdir': False,
-                              'mode': 0o700,
-                              'uid': 100})
+        self.assertEqual(shadow,
+                         {'content': "root:$6$1234567890abcdef$bmJYd0" +
+                                     "uAmObfMzvLnkaaUyiDY4bzdohP0WccC" +
+                                     "pH2NOirA..4nNprFh317qNFrHvVit.n" +
+                                     "ddQx65BSDG/oVouG2." +
+                                     ":14917:0:99999:7:::\n" +
+                                     "bin:*:14495:0:99999:7:::\n" +
+                                     "daemon:*:14495:0:99999:7:::\n",
+                          'gid': 100,
+                          'isdir': False,
+                          'mode': 0o700,
+                          'uid': 100})
         vfs.teardown()
 
     def test_inject_files_into_fs(self):
Index: nova/nova/virt/disk/api.py
===================================================================
--- nova.orig/nova/virt/disk/api.py
+++ nova/nova/virt/disk/api.py
@@ -23,6 +23,8 @@ Includes injection of SSH PGP keys into
 
 """
 
+import ctypes
+import ctypes.util
 import os
 import random
 import tempfile
@@ -41,9 +43,6 @@ from nova.virt.disk.vfs import api as vf
 from nova.virt.image import model as imgmodel
 from nova.virt import images
 
-if os.name != 'nt':
-    import crypt
-
 
 LOG = logging.getLogger(__name__)
 
@@ -607,6 +606,29 @@ def _generate_salt():
     return ''.join([random.choice(salt_set) for c in salt])
 
 
+# This is to be removed whenever
+# https://review.opendev.org/c/openstack/oslo.utils/+/931899
+# is merged.
+if ctypes.util.find_library("crypt"):
+    _libcrypt = ctypes.CDLL(ctypes.util.find_library("crypt"), use_errno=True)
+    _crypt = _libcrypt.crypt
+    _crypt.argtypes = (ctypes.c_char_p, ctypes.c_char_p)
+    _crypt.restype = ctypes.c_char_p
+else:
+    _crypt = None
+
+
+def _crypt_password(key, salt):
+    """Encrtpt password string and generate the value in /etc/shadow format
+    This is provided as a replacement of crypt.crypt method because crypt
+    module was removed in Python 3.13.
+    .. versionadded:: 7.5
+    """
+    if _crypt is None:
+        raise RuntimeError('libcrypt is not available')
+    return _crypt(key.encode('utf-8'), salt.encode('utf-8')).decode('utf-8')
+
+
 def _set_passwd(username, admin_passwd, passwd_data, shadow_data):
     """set the password for username to admin_passwd
 
@@ -621,24 +643,16 @@ def _set_passwd(username, admin_passwd,
     :raises: exception.NovaException(), IOError()
 
     """
-    if os.name == 'nt':
-        raise exception.NovaException(_('Not implemented on Windows'))
 
     # encryption algo - id pairs for crypt()
     algos = {'SHA-512': '$6$', 'SHA-256': '$5$', 'MD5': '$1$', 'DES': ''}
 
     salt = _generate_salt()
 
-    # crypt() depends on the underlying libc, and may not support all
-    # forms of hash. We try md5 first. If we get only 13 characters back,
-    # then the underlying crypt() didn't understand the '$n$salt' magic,
-    # so we fall back to DES.
-    # md5 is the default because it's widely supported. Although the
-    # local crypt() might support stronger SHA, the target instance
-    # might not.
-    encrypted_passwd = crypt.crypt(admin_passwd, algos['MD5'] + salt)
-    if len(encrypted_passwd) == 13:
-        encrypted_passwd = crypt.crypt(admin_passwd, algos['DES'] + salt)
+    # Let's use SHA-512, as all other algo are to be considered insecure.
+    # Note that this drop supports for any distro that still lags behind
+    # and hasn't upgraded its crypt support.
+    encrypted_passwd = _crypt_password(admin_passwd, algos['SHA-512'] + salt)
 
     p_file = passwd_data.split("\n")
     s_file = shadow_data.split("\n")
Index: nova/releasenotes/notes/inject-password-now-uses-sha512-72dde468b662372f.yaml
===================================================================
--- /dev/null
+++ nova/releasenotes/notes/inject-password-now-uses-sha512-72dde468b662372f.yaml
@@ -0,0 +1,6 @@
+---
+features:
+  - |
+    The old inject admin password feature has switched from MD5 to SHA-512, so
+    old Linux distribution that do not understand this hashing algo will not
+    be supported anymore.
